[Proposal][Feature] Convert DMM's ContentID to standard ID

seeyabye opened the pull request:
The current `ID` is rather inconsistent with all other web providers. For instance, DMM uses a format of `ID#####` where # represents some kind of numeric value while others like Javlibrary uses the standard `ID-###`. Additionally, there's also no way to customize which type of ID I would like to use.

This exhibits a problem because I would prefer to have it use `ID-###` rather than `ID####`, but there is no way to reflect this choice in the settings file.

In this PR, I would like to propose to standardize the `ID` format to `ID-###` and keep `ID#####` as a separate ID, called `ContentID`. This would also allow customization in `sort.format.file` and others to support both `<ID>` and `<CONTID>`.

There are two ways to go about this problem.

1. Include a new field in `$Data` to support `ContentId` and `Id`. Possibly, include another field that keeps the original scraped id, called `ScrapeId`
2. Keep the original data structure, but make modifications to `Convert-JVString` to support `<CONTID>`.

Each of the above choices has its own repercussions:

1. Requires refactoring all aggregators to include the additional field for support
2. It's more tricky to maintain the long run because it's the responsibility of `Convert-JVString` to cover all use cases, especially when new aggregators are added. At the moment, I don't see a large deviation in ID patterns.

Right now, I think either choice is fine, but I will leave the final decision to you.

What do you think?

Personally, I have gone with the first choice, but if need be, I can implement the second choice.

With that, I've only made changes to DMM as a prototype, but if this proposal is accepted, I will continue to make changes for the others depending on the choice.

Checklist:
- [x] Implement `Id` and `ContentId` for DMM.
- [ ] ~~Implement `ContentId` for all other aggregators~~

jvlflame commented:
This is a good idea. How I originally envisioned users using Javinizer would be using scrapers in conjunction with each other to grab the necessary metadata fields (i.e. dmm for description, r18 for id, etc.). Because Dmm didn't natively support ID, I didn't include it and assumed a user would just use a different scraper to grab the ID. But you bring up a good point of standardizing the ID format to match the rest of the scrapers.

> There are two ways to go about this problem.
>
> 1. Include a new field in $Data to support ContentId and Id. Possibly, include another field that keeps the original scraped id, called ScrapeId
> 2. Keep the original data structure, but make modifications to Convert-JVString to support <CONTID>.

Definitely option 1 like you already implemented. This is similar to how the R18 scraper is already working.

> With that, I've only made changes to DMM as a prototype, but if this proposal is accepted, I will continue to make changes for the others depending on the choice.

While I think it's feasible to convert from ContentId => Id, I don't believe the other way around is possible, at least not in the original format that dmm uses with things like `h_###aaa00123`. I believe most people will stick with the standardized DVD Id for their naming, and if they do want to use ContentId, they would still be able to with the Dmm and R18 scrapers.

Let's just:
- Add the Id to the Dmm scraper without adding ContentId to the others
- Add setting to prefer the ID type
  - No fallback to the alternate ID type if the preferred is unavailable
- Update `Convert-JVString` to support `<CONTENTID>` <- Let's use the full name for the tag rather than CONTID

jvlflame commented on src/Javinizer/Private/Scraper.Dmm.ps1:
```diff
@@ -14,6 +14,28 @@ function Get-DmmContentId {
+            # Expects ###ID##### or ID#####
+            $contentId = Get-DmmContentId $Webrequest
+            $Id = $contentId
+            $m = ($contentId | Select-String -Pattern '\d*([a-z]+)(\d+)$' -AllMatches).Matches
+            if($m.Groups.Count -gt 2 -and $m.Groups[1] -and $m.Groups[2]) {
+                $Id = $m.Groups[1].Value.ToUpper() + "-" + ($m.Groups[2].Value -replace '^0{1,5}', '').PadLeft(3, '0')
+            }
```
The regex string `\d*([a-z]+)(\d+)$` needs to be updated to include ContentId's that include suffixes.
I'd say just add a `(.*)` to the end to match the suffix and add it to `$Id` in line 30. 

So new regex string of `\d*([a-z]+)(\d+)(.*)$`.

jvlflame replied:
Also, can you add a space in the between the `if` and the parentheses in line 29?

seeyabye commented:
I have presently made the following changes:

- Added Id Preference setting to the configuration file. Users can choose to have `id` or `contentid` for DMM and R18. By default, it is set to `id` which corresponds to the `ID-###` format. 
- I have also included `<CONTENTID>`
- Made changes to DMM Scraper so that `age_check_done` is presented to both DMM-Ja and DMM-En. I found times when DMM-Ja is forcing me to do the age verification before this change.
- Fixed a bug related to label, maker, and series not returning any results for DMM-Ja

**Remark**
I've currently embedded the logic for choosing `ID` or `ContentId` in R18 and DMM's scraper. This means `$movieDataObject.Id` will return `ID-###` or `ID#####` depending on the preference setting configured.

If there's any other preferred way to do this, please let me know.

jvlflame commented on src/Javinizer/Public/Get-JVData.ps1:
```diff
@@ -226,10 +236,10 @@ function Get-JVData {
                     Start-ThreadJob -Name "jvdata-Dmm" -ThrottleLimit $throttleLimit -ScriptBlock {
                         Import-Module $using:jvModulePath
```
Can you change the threadjob name to `jvdata-DmmJa`? Just realized this is inconsistent with the others.

jvlflame commented on src/Javinizer/Public/Get-JVData.ps1:
```diff
@@ -58,6 +58,14 @@ function Get-JVData {
         [Alias('scraper.movie.dmm.scrapeactress')]
         [Boolean]$DmmScrapeActress,
```
The parameter aliases seem to be incorrect here.

`scraper.movie.dmm.scrapeactress` => `scraper.option.dmm.scrapeactress` (This one is my mistake)
`scraper.movie.dmm.idpreference` => `scraper.option.dmm.idpreference`
`scraper.movie.dmm.r18prfeference` => `scraper.option.r18.idpreference`

seeyabye replied:
I have consolidated all of them in `scraper.option.idpreference`.

jvlflame commented on src/Javinizer/Public/Get-DmmData.ps1:
```diff
@@ -45,7 +48,7 @@ function Get-DmmData {
             Source        = if ($Url -match '/en/') { 'dmm' } else { 'dmmja' }
             Url           = $Url
-            Id            = Get-DmmId -WebRequest $webRequest
+            Id            = if ($IdPreference -eq "id") { Get-DmmId -WebRequest $webRequest } elseif ($IdPreference -eq "contentid") { Get-DmmContentId -WebRequest $webRequest }
```
I did a bit of thinking about having the Id convert to ContentId directly on the scrapers themselves.

I think it would be better to change it only on the aggregated data object, since if people were using Javinizer solely for the scraper functions, they wouldn't want the original scraper output to be changed depending on an Id preference.

My approach would be to consolidate the settings into a single setting `scraper.option.idpreference`, remove `ContentId` from the aggregated data object, and set the `Id` during the foreach loop that assigns all the aggregated data fields.
```powershell
foreach ($field in $metadataFields) {
    $metadataPriority = (Get-Variable -Name "$($field)Priority" -ValueOnly)
    foreach ($priority in $metadataPriority) {
        $sourceData = $Data | Where-Object { $_.Source -eq $priority }

        if ($null -eq $aggregatedDataObject.$field) {
            if ($field -eq 'AlternateTitle') {
                $aggregatedDataObject.$field = $sourceData.Title
            } elseif ($field -eq 'Id') {
                if ($IdPreference -eq 'contentid') {
                    $aggregatedDataObject.$field = $sourceData.ContentId
                } else {
                    $aggregatedDataObject.$field = $sourcedata.Id
                }
            } else {
                $aggregatedDataObject.$field = $sourceData.$field
            }
            Write-JVLog -Write:$script:JVLogWrite -LogPath $script:JVLogPath -WriteLevel $script:JVLogWriteLevel -Level Debug -Message "[$($Data[0].Id)] [$($MyInvocation.MyCommand.Name)] [$field - $priority] Set to [$($sourceData.$field | ConvertTo-Json -Compress)]"
        }
    }
}
```

What do you think of this approach?

seeyabye replied:
I personally like this approach.

To be honest, I wasn't quite pleased with what I have done. I prefer not to have the logic encoded in the model (in this case `$movieDataObject`). I will include this in the changes.

jvlflame reviewed (approved):
Looks good!
